src/emqx_channel.erl

```diff
@@ -1794,18 +1813,32 @@ shutdown(Reason, Reply, Packet, Channel) ->
     {shutdown, Reason, Reply, Packet, Channel}.
 
 disconnect_and_shutdown(Reason, Reply, Channel = ?IS_MQTT_V5
-                                               = #channel{conn_state = connected}) ->
-    shutdown(Reason, Reply, ?DISCONNECT_PACKET(reason_code(Reason)), Channel);
+                                               = #channel{conn_state = ConnState}) ->
+    NChannel = ensure_disconnected(ConnState, Reason, Channel),
+    shutdown(Reason, Reply, ?DISCONNECT_PACKET(reason_code(Reason)), NChannel);
```

Comment:
You have changed `conn_state = connected` (a pattern match) to `conn_state = ConnState` (a variable bind).
As a result, it will try to send DISCONNECT_PACKET at disconnected state.